src/stashcache.py

```diff
@@ -140,3 +140,139 @@ def generate_public_authfile(vo_data, legacy=True):
 
     return authfile[:-2]
+
+def _origin_is_allowed(origin_hostname, vo_name, stashcache_data, resource_groups, suppress_errors=True):
+    origin_resource = None
+    for group in resource_groups:
+        for resource in group.resources:
+            if origin_hostname.lower() == resource.fqdn.lower():
+                origin_resource = resource
+                break
+    if not origin_resource:
+        if suppress_errors:
+            return False
+        else:
+            raise ValueError("FQDN {} is not a registered service.".format(origin_hostname))
+    if 'XRootD origin server' not in origin_resource.service_names:
+        if suppress_errors:
+            return False
+        else:
+            raise ValueError("FQDN {} (resource name {}) does not provide an origin service.".format(origin_hostname, origin_resource.name))
+    allowed_vos = origin_resource.data.get("AllowedVOs")
+    if allowed_vos is None:
+        if suppress_errors:
+            return False
+        else:
+            raise ValueError("Origin server at {} (resource name {}) does not provide an AllowedVOs list.".format(origin_hostname, origin_resource.name))
+
+    matches_origin = False
+    for vo in allowed_vos:
+        if vo == 'ANY' or vo == vo_name:
+            matches_origin = True
+            break
+    if not matches_origin:
+        return False
+
+    allowed_origins = stashcache_data.get("AllowedOrigins")
+    if allowed_origins is None:
+        if suppress_errors:
+            return False
+        else:
+            raise ValueError("VO {} in StashCache does not provide an AllowedOrigins list.".format(vo_name))
+    for origin_name in allowed_origins:
+        if origin_name == origin_resource.name:
+            return True
+    return False
+
+def _get_allowed_caches(stashcache_data, resource_groups, suppress_errors=True):
+    allowed_caches = stashcache_data.get("AllowedCaches")
+    if allowed_caches is None:
+        if suppress_errors:
+            return []
+        else:
+            raise ValueError("VO {} enables StashCache but does not specify the allowed caches.".format(vo_name))
+
+    resources = []
+    for group in resource_groups:
+        for resource in group.resources:
+            # First, does this provide a cache service?
+            if 'XRootD cache server' not in resource.service_names:
+                continue
+
+            # Next, does it allow this VO?  Unlike the StashCache origin case requiring the origin to list AllowedVOs,
+            # we do not consider the lack of AllowedVOs an error as the cache doesn't
+            # explicitly record *which* data federation it is participating in (might not be SC!).
+            matches_vo = False
+            for vo in resource.data.get("AllowedVOs", []):
+                if vo == 'ANY':
+                    matches_vo = True
+                    break
+                elif vo == 'PUBLIC':
+                    continue
+                elif vo == vo_name:
```

Comment:
`vo_name` is undefined in several places in this function. Where is it supposed to come from?

Reply:
I'll fix this in a PR to your PR.